components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java

```diff
@@ -75,84 +75,135 @@ public static AuthorizationHandlerManager getInstance() throws IdentityOAuth2Exc
     public OAuth2AuthorizeRespDTO handleAuthorization(OAuth2AuthorizeReqDTO authzReqDTO)
             throws IdentityOAuth2Exception, IdentityOAuthAdminException, InvalidOAuthClientException {
 
-        String responseType = authzReqDTO.getResponseType();
-        OAuth2AuthorizeRespDTO authorizeRespDTO = new OAuth2AuthorizeRespDTO();
-
-        if (!responseHandlers.containsKey(responseType)) {
-            log.warn("Unsupported Response Type : " + responseType +
-                    " provided  for user : " + authzReqDTO.getUser());
-            handleErrorRequest(authorizeRespDTO, OAuthError.CodeResponse.UNSUPPORTED_RESPONSE_TYPE,
-                    "Unsupported Response Type!");
-            authorizeRespDTO.setCallbackURI(authzReqDTO.getCallbackUrl());
+        OAuthAuthzReqMessageContext authzReqMsgCtx = getoAuthAuthzReqMessageContext(authzReqDTO);
+        ResponseTypeHandler authzHandler = responseHandlers.get(authzReqDTO.getResponseType());
+        OAuth2AuthorizeRespDTO authorizeRespDTO = validateAuthzRequest(authzReqDTO, authzReqMsgCtx, authzHandler);
+        if (foundErrorResponse(authorizeRespDTO)) {
+            if (log.isDebugEnabled()) {
+                log.debug("Error response received for authorization request by user : " + authzReqDTO.getUser()  +
+                        ", client : " + authzReqDTO.getConsumerKey() + ", scope : " +
+                        OAuth2Util.buildScopeString(authzReqDTO.getScopes()));
+            }
             return authorizeRespDTO;
         }
+        try {
+	    // set the authorization request context to be used by downstream handlers. This is introduced as a fix for
+	    // IDENTITY-4111
+	    OAuth2Util.setAuthzRequestContext(authzReqMsgCtx);
```

Comment:
Couldn't find any place which uses this threadLocal variable. Shall we more analyze this is required or not

Reply:
We give the support to extend the org.apache.oltu.oauth2.as.issuer.OAuthIssuer implementation in OauthTokenIssuerImpl (check [1]). org.apache.oltu.oauth2.as.issuer.OAuthIssuer APIs do not accept any messageContext data. authzHandler.issue() method may call the implemented OAuthIssuer APIs in downstream. In that case, OAuthIssuer APIs need the threadLocal to retrieve authzReqMsgCtx data. Therefore cannot remove this.

[1] https://github.com/wso2-extensions/identity-inbound-auth-oauth/blob/master/components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/OauthTokenIssuerImpl.java#L29